api/api/utils/photon.py

```diff
@@ -142,53 +145,34 @@ def get(
             params=params,
             headers=headers,
         )
-        upstream_response.raise_for_status()
-    except requests.ReadTimeout as exc:
-        # Count the incident so that we can identify providers with most timeouts.
-        key = f"{settings.THUMBNAIL_TIMEOUT_PREFIX}{domain}"
-        try:
-            cache.incr(key)
-        except ValueError:  # Key does not exist.
-            cache.set(key, 1)
-
-        sentry_sdk.capture_exception(exc)
-        raise UpstreamThumbnailException(
-            f"Failed to render thumbnail due to timeout: {exc}"
+        cache.incr(f"thumbnail_response_code:{month}:{upstream_response.status_code}")
+        cache.incr(
+            f"thumbnail_response_code_by_domain:{domain}:"
+            f"{month}:{upstream_response.status_code}"
         )
-    except requests.HTTPError as exc:
-        # Save number of occurrences to redis. Only raise a Sentry error when
-        # the number of occurrences reaches the configured threshold, to
-        # prevent provider API downtime from causing excessive Sentry events.
-        key = f"{settings.THUMBNAIL_HTTP_ERROR_PREFIX}{domain}"
-        errors = 1
-        try:
-            errors = cache.incr(key)
-        except ValueError:  # Key does not exist.
-            cache.set(key, 1)
-        if errors >= settings.THUMBNAIL_HTTP_ERROR_THRESHOLD_TO_NOTIFY and (
-            errors % settings.THUMBNAIL_HTTP_ERROR_THRESHOLD_FOR_REPEATED_NOTIFICATIONS
-            == 0
+        upstream_response.raise_for_status()
+    except Exception as exc:
+        exception_name = f"{exc.__class__.__module__}.{exc.__class__.__name__}"
+        key = f"thumbnail_error:{exception_name}:{domain}:{month}"
+        count = cache.incr(key)
+        if count <= settings.THUMBNAIL_ERROR_INITIAL_ALERT_THRESHOLD or (
+            count % settings.THUMBNAIL_ERROR_REPEATED_ALERT_THRESHOLD == 0
```

Comment:
A misunderstanding of the issue description is that we send _every thousandth error to Sentry_ but in reality, this was limiting only HTTP errors, which were the main source of our Sentry events consumption. Every other error type was sent and I think this was the main motivation for the issue. Should we keep the same limit narrowed here?

Reply:
I think this strategy is worth having in place for all exceptions - it's possible that some other issue might arise which is _not_ an HTTP error but still gets raised in the thousands and could make us hit our quota. By specifying a minimum threshold we're still allowing those initial 100 events through (which currently almost all of the alerts in question have <100 besides the HTTP error one) while ensuring Sentry doesn't get overwhelmed.

Especially given that this essentially resets every month and is per domain, I think it's reasonable to keep the threshold logic here for everything.